### ISSUE TITLE
`systemctl docker start` get stuck in cloud-init

### ISSUE THREAD

**Issue by zigarn** [moby/moby#41767]
**Description**

Whith Docker 20.10 on CentOS 7, when starting it in a cloud-init script, the start is stuck.

**Steps to reproduce the issue:**
1. Create a cloud-init script that install and start docker
2. Launch a CentOS 7 VM with this cloud-init
3. cloud-init script is stuck on startin docker

**Describe the results you received:**

`systemctl docker start` hangs forever.

**Describe the results you expected:**

Docker daemon starts and cloud-init script can continue.

**Additional information you deem important (e.g. issue happens only occasionally):**

Looks related to #41297.
If I remove `multi-user.target` from after and launch a `systemctl daemon-reload`, the `systemctl start docker` get unstucked.

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           20.10.0
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        7287ab3
 Built:             Tue Dec  8 18:57:35 2020
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.0
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       eeddea2
  Built:            Tue Dec  8 18:56:55 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.4.2-docker)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 0
 Server Version: 20.10.0
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc io.containerd.runc.v2 io.containerd.runtime.v1.linux
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc version: ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 3.10.0-693.11.6.el7.x86_64
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 3.701GiB
 Name: ip-172-31-21-73.eu-central-1.compute.internal
 ID: 2IU4:K3UK:R5X3:2FVQ:JILR:F5US:GLDG:5JC4:KG5F:RGAI:UVBS:3ZIZ
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

AWS VM Centos7 (AMI: ami-337be65c on eu-central-1 I think)

**Comment by yuxhuang** [comment]
I experience the same lock up on DigitalOcean Ubuntu focal. The machines are auto updated every couple days then hit this issue. These machines also have a `rc.local` to start docker containers. `dockerd` successfully starts when I manually kill the `docker start` commands.

Pinning my versions back to `5:19.03.14~3-0~ubuntu-focal` solves the issue.

**Comment by thaJeztah** [comment]
Thanks for reporting; sounds indeed related to https://github.com/moby/moby/pull/41297 (and https://github.com/docker/docker-ce-packaging/pull/488) from your description

**Comment by thaJeztah** [comment]
I found https://bugzilla.redhat.com/show_bug.cgi?id=1428492 (duplicate of https://bugzilla.redhat.com/show_bug.cgi?id=1400249), which could be related. That mentions `cloud-init-0.7.9-4.fc26` and `cloud-init-0.7.8-6.fc25` to fix this (from a quick scan), but obviously those are Fedora packages, so not sure if similar fixes exist for CentOS 7

**Comment by thaJeztah** [comment]
> I experience the same lock up on DigitalOcean Ubuntu focal. The machines are auto updated every couple days then hit this issue. These machines also have a rc.local to start docker containers. dockerd successfully starts when I manually kill the docker start commands.

Sorry, just saw you had the issue on a different situation.

I'm wondering if that particular case could be addressed by using systemd instead of a `rc.local` file (https://unix.stackexchange.com/a/479766/78656); I can imagine there being various corner cases when combining the two. Using a systemd unit would likely better declare a dependency between your script and the docker service being available.

**Comment by yuxhuang** [comment]
> > I experience the same lock up on DigitalOcean Ubuntu focal. The machines are auto updated every couple days then hit this issue. These machines also have a rc.local to start docker containers. dockerd successfully starts when I manually kill the docker start commands.
> 
> Sorry, just saw you had the issue on a different situation.
> 
> I'm wondering if that particular case could be addressed by using systemd instead of a `rc.local` file (https://unix.stackexchange.com/a/479766/78656); I can imagine there being various corner cases when combining the two. Using a systemd unit would likely better declare a dependency between your script and the docker service being available.

That's fair. I think `systemd`'s getting a bit invasive here when the old expectation is `rc.local` runs at the end of everything. I'll probably change it to a system unit.

**Comment by thaJeztah** [comment]
Yeah, systemd is often "all or nothing"

**Comment by rwky** [comment]
I've come across something similar which is related to #41297 This is on Ubuntu 20.04 with docker-ce 20.10

If a systemd unit relies on docker and is set to install on multi-user.target it will never start

Sample systemd file
```
[Unit]
Requires=docker.service
After=docker.service

[Service]
ExecStart=/usr/bin/docker run nginx

[Install]
WantedBy=multi-user.target
```

The problem is that docker is set to start after `multi-user.target`

The docker systemd unit has:
    
`WantedBy=multi-user.target`

Which tells systemd that when multi-user.target starts docker.service should start

It also has:
`After=network-online.target firewalld.service containerd.service multi-user.target`

Which tells systemd that docker should start after multi-user.target has fully started

These are conflicting we're saying "multi-user.service depends on docker but docker must run after multi-user.service"

In the case of the sample unit it won't start because it starts when multi-user.target starts but can't because it required docker to be running which won't start until after multi-user.target starts.

I suggest reverting 0ca7456e5284d4aa9f3e37e69c7c93eff4420d3d I can't think of a better solution at the moment and if you search the internet for `start docker container with systemd` you'll find a ton of unit files which look like the sample so this has the potential to break things for a lot of people.

**Comment by rwky** [comment]
Note to anyone wanting to override the unit, the whole unit needs to be copied you can't use a drop-in, from the systemd manual

>  Dependencies (After=, etc.) cannot be reset to an empty list, so dependencies can only be added in drop-ins. If you want to remove dependencies, you have to override the entire unit.

**Comment by bwinterton** [comment]
I can confirm that the circular dependency pointed out by @rwky is what was causing my issues. Attempting to run Docker 20.10 with systemd resulted in the service never starting. Removing `multi-user.target` from the `After=` allowed the service to start up as expected.

**Comment by thaJeztah** [comment]
> If a systemd unit relies on docker and is set to install on multi-user.target it will never start

is this a unit you defined yourself, or is that part of some package? (curious what package the conflict is in)

> Attempting to run Docker 20.10 with systemd resulted in the service never starting

@bwinterton is this only in the case of cloud-init / automated install, or also when manually installed / started?

I recall looking up if `Wants=` / `WantedBy=` and `After=` would conflict https://github.com/moby/moby/pull/41297#issuecomment-665926970, which gave me this; https://unix.stackexchange.com/a/547662/78656 (obviously stack exchange is not always a trusted source, so may be wrong);

> `After` does not imply `Wants` or `WantedBy`, nor does it conflict with those settings

Also looking at the [systemd Before=, After= section](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Before=)

> Note that those settings are independent of and orthogonal to the requirement dependencies as configured by Requires=, Wants=, Requisite=, or BindsTo=. It is a common pattern to include a unit name in both the After= and Wants= options, in which case the unit listed will be started before the unit that is configured with these options.

Admitted, that mentions `Wants=`, not `WantedBy=`

I see @tianon left a comment about this; https://github.com/moby/moby/pull/41297#issuecomment-665972844

**Comment by rwky** [comment]
>is this a unit you defined yourself, or is that part of some package? (curious what package the conflict is in)

One I wrote the one I supplied is the simplest form. You'll find lots of similar ones on the internet as a guide.

The wants/wanted by/after etc. relationship is complicated. 

The gist is that after determines the order where as wants determines the dependencies. This is why it's suggested to use both. 

So for example if you want a service to only start once docker is ready you use

```
Requires=docker.service
After=docker.service
```

The Require means start docker when the services starts. The After means start the service once docker has fully started.

Wants is just a less strict version of requires (it doesn't cause failure)

So WantedBy and After of the same service doesn't make a lot of sense.

**Comment by calvincheng8** [comment]
I have encountered the same circular dependencies problem described by @rwky. 

docker.service is part of multi-user.target and also only starts after multi-user.target. This does not make sense.

**Comment by calvincheng8** [comment]
20.10 changed the following line in docker.service.
```
After=network-online.target firewalld.service containerd.service multi-user.target
```
The last entry was added in 20.10.

After I removed `multi-user.target` from the `After=...`, both docker.service and the docker container defined in another service unit started without problems. See  docker/for-linux#1162

**Comment by pvizeli** [comment]
As my issue was closed. The base of my Issue was:
- Docker doing it now wrong: https://www.freedesktop.org/software/systemd/man/bootup.html
- If someone run wifi and needs a hack for fast boot, they can overwrite it locally with the `docker.socket.d` folder

**Comment by gcstang** [comment]
Ran into this issue or very similar with CentOS 7.7 as well systemd won't start docker, had to perform yum history undo to go back to 
docker-ce 19.03.14-3.el7
docker-ce-cli 19.03.14-3.el7
containerd.io 1.3.9-3.1.el7

Upgrade was
docker-ce 20.10.0-3.el7
docker-ce-cli.x86_64 1:20.10.0-3.el7
docker-ce-rootless-extras.x86_64 0:20.10.0-3.el7
containerd.io.x86_64 0:1.4.3-3.1.el7

**Comment by GuyBachkhaznadji** [comment]
We added this to our start up scripts for now and that seems to have fixed it:
`sed -i 's/containerd.service multi-user.target/containerd.service/g' /usr/lib/systemd/system/docker.service `

**Comment by zigarn** [comment]
> We added this to our start up scripts for now and that seems to have fixed it:
> `sed -i 's/containerd.service multi-user.target/containerd.service/g' /usr/lib/systemd/system/docker.service `
> 
> _**Edit**_
> We're using CentOS 7

Using this on my side:

```shell
sed --in-place 's/^\(After=.*\) multi-user.target\(.*\)/\1\2/' /usr/lib/systemd/system/docker.service
```

**Comment by bwinterton** [comment]
>>Attempting to run Docker 20.10 with systemd resulted in the service never starting

>@bwinterton is this only in the case of cloud-init / automated install, or also when manually installed / started?

@thaJeztah My use case is from a manual installation on Raspberry Pi OS (Debian Buster) from the convenience script and then updated to 20.10 through `apt upgrade`

**Comment by tiborvass** [comment]
@zigarn  or anyone hitting this issue would you mind running:

```
curl -fsSL https://get.docker.com/ | DOWNLOAD_URL=https://download-stage.docker.com REPO_FILE=docker-ce-staging.repo sh
```

And confirm you no longer see the issue?

**Comment by zigarn** [comment]
@tiborvass I confirm that the issue is gone with staging package.

**Comment by rwky** [comment]
I can confirm the staging package works with my use case as well.

**Comment by raisinbrandon** [comment]
I have the same issue as rwky, and staging has resolved it.

**Comment by thaJeztah** [comment]
The docker 20.10.1 packages are now available on download.docker.com; make sure to switch back repositories to the main (download.docker.com) package repo if you temporarily switched to the staging repository.

I'm closing this issue, as this should be resolved, but feel free to continue the conversation.

**Comment by raisinbrandon** [comment]
Thanks for the quick turnaround, Sebastiaan!

I noticed the docker binary did not change, so it still reports version 20.10.0.  The package version correctly indicates 20.10.1, though.  Perhaps that was intentional, so I'm not going to file a bug report.  I'm mainly pointing it out for anyone reading this thread.

**Comment by thaJeztah** [comment]
@raisinbrandon could it be it's still installing from the staging server? I just tried upgrading a server, and it does show 20.10.1;

```
docker version
Client: Docker Engine - Community
 Version:           20.10.1
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        831ebea
 Built:             Tue Dec 15 04:34:59 2020
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.1
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       f001486
  Built:            Tue Dec 15 04:32:40 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Comment by raisinbrandon** [comment]
@thaJeztah i figured out the issue, and it was my mistake.  i did indeed make sure i was no longer pulling from staging before i posted my comment.

however, when i said the package was showing 20.10.1, i was looking at docker-ce.x86_64 from docker-ce-stable, but i completely failed to notice that docker-ce-cli.x86_64 still showed 20.10.0.  it seems that i need to poke around in my ansible scripts and figure out why that package didn't get upgraded.

i really appreciate your response, and i apologize for the false alarm.  thanks again for the fix!

**Comment by thaJeztah** [comment]
Good to hear it's resolved, and no apologies needed; various parameters have to be set (and passed through) our release pipeline, so it would've been possible that a parameter didn't pass through somewhere 😬😬

As to your scripts: when doing a fresh install of "only" the "docker-ce" package, the deb/rpm "depends/requires" settings of that package will automatically install the latest versions of "docker-ce-cli" and "containerd.io". 
When *updating*, those dependencies are not updated automatically, so to update all; make sure to update all of "docker-ce", "docker-ce-cli" and "containerd.io" to get their latest (patch) releases.

That said; for production setups, it may be recommended to explicitly set/pin versions to install (after testing/verifying new packages in staging) to prevent unexpected upgrades that have not been verified in your environment.

**Comment by raisinbrandon** [comment]
thanks for the tips.  you confirmed what i've just found regarding the first time install and updates.  we're using an ansible role by geerlingguy to deploy docker, and it only contains docker-ce.  one of my colleagues configured it, so i'll have to poke around and see if there's an easy way to have it update the other packages as well.

i'm also in complete agreement with you about pinning versions, although we would still have this same issue anytime we wanted to upgrade the pinned version through ansible.  i believe my colleague had a specific reason for not doing pinning a version, but i'll revisit the issue with him after the holidays.

thanks again for the advice.

**Comment by boeboe** [comment]
```console
$ cat /etc/systemd/system/docker.service.d/override.conf
# Disable flags to dockerd, all settings are done in /etc/docker/daemon.json
[Service]
ExecStart=
ExecStart=/usr/bin/dockerd
```
